# Ejercicio 7. Pandas

#### Al. Daniela Lizeth Molina Barrera

1. Utiliza 15 funciones de pandas para manipular datos. Asegúrate de incluir funciones de agregación y de agrupamiento. Comenta la manipulación del dataset a cada paso y explica por qué utilizas esa función o método en particular. Sube tu notebook.

In [2]:
%conda install pandas

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.10.3
  latest version: 4.11.0

Please update conda by running

    $ conda update -n base conda



# All requested packages already installed.


Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as pd 

In [4]:
# Funcion 1. Leer un csv con ayuda de pandas
costumers = pd.read_csv('customers.csv')
costumers

customer_id  year_of_birth  educational_level marital_status  \
0       20201701            1982        Graduation         Single   
1       20201702            1950        Graduation        Married   
2       20201703            1965        Graduation       Divorced   
3       20201704            1984        Graduation   Relationship   
4       20201705            1981               PhD        Widowed   
..           ...             ...               ...            ...   
494     20202195            1944               PhD       Divorced   
495     20202196            1962            Master       Divorced   
496     20202197            1978        Graduation       Divorced   
497     20202198            1971               PhD   Relationship   
498     20202199            1949               PhD   Relationship   

     annual_income purhcase_date  recency  online_purchases  store_purchases  \
0          58138.0      9/4/2012       58                 8                4   
1          46344.0      3/8/2014       38                 1                2   
2          71613.0     8/21/2013       26                 8               10   
3          26646.0     2/10/2014       26                 2                4   
4          58293.0     1/19/2014       94                 5                6   
..             ...           ...      ...               ...              ...   
494        55614.0    11/27/2013       85                 9                6   
495        59432.0     4/13/2013       88                 5               11   
496        55563.0      4/5/2014       22                 2                3   
497        43624.0     4/21/2013       83                 4                4   
498        41461.0     5/22/2014       63                 6               11   

     complaints  calls  intercoms  
0             0      3         11  
1             0      3         11  
2             0      3         11  
3             0      3         11  
4             0      3         11  
..          ...    ...        ...  
494           0      3         11  
495           0      3         11  
496           0      3         11  
497           0      6         11  
498           0      6         11  

[499 rows x 12 columns]

In [5]:
# Funcion 2. Es importante saber la información del data frame que se utilizara 
costumers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499 entries, 0 to 498
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        499 non-null    int64  
 1   year_of_birth      499 non-null    int64  
 2   educational_level  499 non-null    object 
 3   marital_status     499 non-null    object 
 4   annual_income      486 non-null    float64
 5   purhcase_date      499 non-null    object 
 6   recency            499 non-null    int64  
 7   online_purchases   499 non-null    int64  
 8   store_purchases    499 non-null    int64  
 9   complaints         499 non-null    int64  
 10  calls              499 non-null    int64  
 11  intercoms          499 non-null    int64  
dtypes: float64(1), int64(8), object(3)
memory usage: 46.9+ KB


In [6]:
# Funcion 3. Devuelve una tupla con el número de filas y columnas del DataFrame
costumers.shape

(499, 12)

In [7]:
# Funcion 4. Devuelve el número de elementos del DataFrame.
costumers.size

5988

In [8]:
# Funcion 5. Devuelve una lista con los nombres de las columnas del DataFrame.
costumers.columns

Index(['customer_id', 'year_of_birth ', 'educational_level', 'marital_status',
       'annual_income', 'purhcase_date', 'recency', 'online_purchases',
       'store_purchases', 'complaints', 'calls', 'intercoms'],
      dtype='object')

Comenzaremos a trabajar con nuestras variables de 'costumers'. Y para ello 
averiguaremos mas acerca de nuestros clientes.

In [50]:
# Funcion 6. Obtener las primeras lineas del DataFrame
costumers.head(5)

customer_id  year_of_birth  educational_level marital_status  \
0     20201701            1982        Graduation         Single   
1     20201702            1950        Graduation        Married   
2     20201703            1965        Graduation       Divorced   
3     20201704            1984        Graduation   Relationship   
4     20201705            1981               PhD        Widowed   

   annual_income purhcase_date  recency  online_purchases  store_purchases  \
0        58138.0      9/4/2012       58                 8                4   
1        46344.0      3/8/2014       38                 1                2   
2        71613.0     8/21/2013       26                 8               10   
3        26646.0     2/10/2014       26                 2                4   
4        58293.0     1/19/2014       94                 5                6   

   complaints  calls  intercoms  
0           0      3         11  
1           0      3         11  
2           0      3         11  
3           0      3         11  
4           0      3         11

In [51]:
# Funcion 7. Obtener las ultimas lineas del DataFrame 
costumers.tail(5)

customer_id  year_of_birth  educational_level marital_status  \
494     20202195            1944               PhD       Divorced   
495     20202196            1962            Master       Divorced   
496     20202197            1978        Graduation       Divorced   
497     20202198            1971               PhD   Relationship   
498     20202199            1949               PhD   Relationship   

     annual_income purhcase_date  recency  online_purchases  store_purchases  \
494        55614.0    11/27/2013       85                 9                6   
495        59432.0     4/13/2013       88                 5               11   
496        55563.0      4/5/2014       22                 2                3   
497        43624.0     4/21/2013       83                 4                4   
498        41461.0     5/22/2014       63                 6               11   

     complaints  calls  intercoms  
494           0      3         11  
495           0      3         11  
496           0      3         11  
497           0      6         11  
498           0      6         11

In [17]:
# Funcion 8. Crearemos un nuevo DataFrame para saber en que año nacieron la mayor
# parte de nuestros clientes.
gen_clientes = costumers.loc[:,['customer_id','year_of_birth ']]
gen_clientes

customer_id  year_of_birth 
0       20201701            1982
1       20201702            1950
2       20201703            1965
3       20201704            1984
4       20201705            1981
..           ...             ...
494     20202195            1944
495     20202196            1962
496     20202197            1978
497     20202198            1971
498     20202199            1949

[499 rows x 2 columns]

In [18]:
# Funcion 9. Ahora usaremos una funcion que agrupe nuestro anterior DataFrame
# de tal forma que nos muestre el año de nacimiento de nuestros clientes.
gen_clientes.groupby('year_of_birth ').agg('count')

customer_id
year_of_birth              
1899                      1
1940                      1
1943                      2
1944                      1
1946                      1
1948                      1
1949                      3
1950                      3
1951                      4
1952                     16
1953                      2
1954                      2
1955                      4
1956                      5
1957                     27
1958                      2
1959                      2
1960                      4
1961                      1
1962                      1
1963                      5
1964                      3
1965                      5
1966                      4
1967                     11
1968                      3
1969                      7
1970                     18
1971                      6
1972                      4
1973                     18
1974                      7
1975                     19
1976                     10
1977                      4
1978                      5
1979                      1
1980                     24
1981                     15
1982                     26
1983                     14
1984                     14
1985                     15
1986                     16
1987                     12
1988                     15
1989                     15
1990                     13
1991                     11
1992                     12
1993                     11
1994                     11
1995                     11
1996                     10
1997                     10
1998                     11
1999                     10
2000                     10

De acuerdo a lo anterior, encontramos que lla mayor cantidad de clientes nacieron en 1957 con 27 clientes (65 años), seguido de 1982 con 26 clientes (40 años), en tercer luegar se encuentra 1981 con 25 clientes (39 años) y en cuarto lugar 1980 (38 años) con 24 clientes. La mayoria de nuestros clientes son mayores de edad y problablemente poseen una vida adulta estable.

Para ello seguiremos investigando mas sobre ellos ...

In [28]:
# Funcion 10. Haremos un slice diferente de nuestros clientes ahora con el nivel de educacion.
level_edu_clientes= costumers[['year_of_birth ','educational_level','annual_income']]
level_edu_clientes

year_of_birth  educational_level  annual_income
0              1982        Graduation        58138.0
1              1950        Graduation        46344.0
2              1965        Graduation        71613.0
3              1984        Graduation        26646.0
4              1981               PhD        58293.0
..              ...               ...            ...
494            1944               PhD        55614.0
495            1962            Master        59432.0
496            1978        Graduation        55563.0
497            1971               PhD        43624.0
498            1949               PhD        41461.0

[499 rows x 3 columns]

In [30]:
# Función 11. Ahora agrupamos por el 'nivel de educacion' y hacemos una agregacion por 'ingreso anual'.
level_edu_clientes.groupby('educational_level').annual_income.agg(['count', 'min', 'max'])

count     min       max
educational_level                         
Basic                  7  7500.0   28249.0
Graduation           249  2447.0  102692.0
High School           39  7500.0   89572.0
Master                80  7500.0   92859.0
PhD                  111  5648.0  157243.0

Encontramos que 249 de nuestros clientes estan graduados de la universidad y como minimo tienen un ingreso anual de 2,447 dolares y un ingreso anual maximo de 102,692 dolares. Seguido de 111 clientes con un PhD (Doctorado) donde el ingreso minimo anual es de 5,648 dolares y un ingreso maximo anual de 157,243 dolares.

Ahora buscaremos saber si tenemos mas ventas en linea o en la tienda...

In [32]:
# Funcion 12. Haremos un tercer slice 
type_of_purchases = costumers[['year_of_birth ','online_purchases','store_purchases']]
type_of_purchases

year_of_birth   online_purchases  store_purchases
0              1982                 8                4
1              1950                 1                2
2              1965                 8               10
3              1984                 2                4
4              1981                 5                6
..              ...               ...              ...
494            1944                 9                6
495            1962                 5               11
496            1978                 2                3
497            1971                 4                4
498            1949                 6               11

[499 rows x 3 columns]

In [46]:
# Funcion 13. Volveremos a agrupar y agregar por el tipo de compras
type_of_purchases.groupby('year_of_birth ').agg({'online_purchases' : ['sum'], 'store_purchases' : ['sum']})

online_purchases store_purchases
                            sum             sum
year_of_birth                                  
1899                          4               4
1940                          2               4
1943                         12              10
1944                          9               6
1946                          4               6
1948                          4               4
1949                         15              25
1950                          5              10
1951                         18              20
1952                         72             103
1953                          5               9
1954                         11              21
1955                         17              35
1956                         36              48
1957                         93             157
1958                          5              15
1959                          4              12
1960                         18              32
1961                          6               7
1962                          5              11
1963                         16              35
1964                         13              29
1965                         22              34
1966                         17              23
1967                         45              74
1968                         28              21
1969                         21              47
1970                         65              98
1971                         24              31
1972                         10              16
1973                         42              90
1974                         36              44
1975                         70             117
1976                         44              44
1977                         20              22
1978                         10              15
1979                          0               0
1980                        117             149
1981                         65              81
1982                        107             185
1983                         52              67
1984                         42              62
1985                         73              83
1986                         95              80
1987                         46              64
1988                         40              72
1989                         59              90
1990                         65              94
1991                         45              60
1992                         37              52
1993                         37              43
1994                         59              72
1995                         38              57
1996                         34              50
1997                         35              81
1998                         42              63
1999                         39              62
2000                         26              38

Nos podemos dar cuenta que de acuerdo a nuestros grueso de clientes que habiamos presentado anteriormente son lo que realizaron la mayor cantidad de compras, sin embargo los del año 1957 son los que hacen la mayoria de sus comprar en linea, a comparacion de los del año 1980 a 1983 los cuales realizan casi la misma cantidad de compras tanto en linea como en tienda.

In [47]:
# Funcion 14. Sumar por columnas, en este caso sumar todas las compras online
total_online = type_of_purchases['online_purchases'].sum()
total_online

1981

In [48]:
# Funcion 15.Sumar por columnas, en este caso sumar todas las compras en tienda 
total_store = type_of_purchases['store_purchases'].sum()
total_store

2884

Y tambien descubrimos que aun la mayoria de nuestras compras siguen siendo realizadas en la tienda con 2884 compras a comparacion de 1981 compras online. Y muy probablemente se deba a las generaciones de nuestros clientes.

TypeError: Can only merge Series or DataFrame objects, a <class 'numpy.int64'> was passed